miri-script/src/commands.rs

```diff
@@ -0,0 +1,637 @@
+        let target_dir = std::env::var_os("CARGO_TARGET_DIR")
+            .filter(|val| !val.is_empty())
+            .unwrap_or_else(|| {
+                let target_dir = path!(runner.miri_dir / "target");
+                target_dir.into()
+            });
+        runner.set_env("CARGO_TARGET_DIR", target_dir);
+
+        // We configure dev builds to not be unusably slow.
+        let devel_opt_level = std::env::var_os("CARGO_PROFILE_DEV_OPT_LEVEL")
+            .filter(|val| !val.is_empty())
+            .unwrap_or_else(|| "2".into());
+        runner.set_env("CARGO_PROFILE_DEV_OPT_LEVEL", devel_opt_level);
+        let rustflags = {
+            let env = std::env::var_os("RUSTFLAGS");
+            let mut flags_with_warnings = OsString::from(
+                "-Zunstable-options -Wrustc::internal -Wrust_2018_idioms -Wunused_lifetimes -Wsemicolon_in_expressions_from_macros ",
+            );
+            if let Some(value) = env {
+                flags_with_warnings.push(value);
+            }
+            // We set the rpath so that Miri finds the private rustc libraries it needs.
+            let mut flags_with_compiler_settings = OsString::from("-C link-args=-Wl,-rpath,");
+            flags_with_compiler_settings.push(&runner.extra_vars.libdir);
+            flags_with_compiler_settings.push(flags_with_warnings);
+            flags_with_compiler_settings
+        };
+        runner.set_env("RUSTFLAGS", rustflags);
+        runner.execute()
+    }
+    fn execute(&mut self) -> Result<()> {
+        // Run command.
+        match self.command {
+            Subcommands::Install { flags } => self.install(flags),
+            Subcommands::Build { flags } => self.build(flags),
+            Subcommands::Check { flags } => self.check(flags),
+            Subcommands::Test { bless, flags } => self.test(*bless, flags),
+            Subcommands::Run { dep, flags } => self.run(*dep, flags),
+            Subcommands::Fmt { flags } => self.fmt(flags),
+            Subcommands::Clippy { flags } => self.clippy(flags),
+            Subcommands::Cargo { flags } => self.cargo(flags),
+            Subcommands::ManySeeds { command, seed_start, seeds } =>
+                self.many_seeds(command, *seed_start, *seeds),
+            Subcommands::Bench { benches } => self.bench(benches),
+            Subcommands::Toolchain { flags } => self.toolchain(flags),
+            Subcommands::RustcPull { commit } => self.rustc_pull(commit.clone()),
+            Subcommands::RustcPush { github_user, branch } => self.rustc_push(github_user, branch),
+        }
+    }
+
+    fn set_env(
+        &mut self,
+        key: impl Into<OsString>,
+        value: impl Into<OsString>,
+    ) -> Option<OsString> {
+        self.env.insert(key.into(), value.into())
+    }
+
+    /// Prepare and set MIRI_SYSROOT. Respects `MIRI_TEST_TARGET` and takes into account
+    /// locally built vs. distributed rustc.
+    fn find_sysroot(&mut self) -> Result<()> {
```

Comment:
Should probably be `find_miri_sysroot` to avoid confusion with `extras.sysroot`.

Same for `build_sysroot` → `build_miri_sysroot`